### PR TITLE
OAuth sources refactor

### PR DESCRIPTION
**Pull request by julsemaan**
## Description

Rework the OAuth sources to add person lookup and to lighten copy paste in the sources.
## Impacts
- Makes the code cleaner and more maintainable
- Adds information to the OAuth registered users in the person table
- Fixes the Github Oauth source, which from what I saw never actually worked
## NEWS file entries

New Features
++++++++++++
- The information available from an OAuth source (first name, last name, ...) are now added to the person when registering

Bug Fixes
++++++++
- Fixed Github authentication source.

### CODE REVIEW

File: html/captive-portal/lib/captiveportal/PacketFence/Controller/Oauth2.pm [hunk 1]
```diff
@@ -193,12 +194,16 @@ sub oauth2Result : Path : Args(1) {
...
                 my $json      = new JSON;
                 my $json_text = $json->decode($response->content());
-                if ($provider eq 'google' || $provider eq 'github' || $provider eq 'facebook') {
+                if ($provider eq 'google' || $provider eq 'facebook') {
                     $pid = $json_text->{email};
                 } elsif ($provider eq 'windowslive'){
                     $pid = $json_text->{emails}->{account};
+                } elsif ($provider eq 'github'){
+                    # github doesn't provide email by default
+                    $pid = $json_text->{login}.'@github';
```
Comment (fdurand) [inline html/captive-portal/lib/captiveportal/PacketFence/Controller/Oauth2.pm:203]: @github.com

Comment (julsemaan) [inline html/captive-portal/lib/captiveportal/PacketFence/Controller/Oauth2.pm:203]: The email address would be invalid (I tested). 
Putting the full domain would suggest that's it's a real e-mail address in my opinion.

File: lib/pf/Authentication/Source/FacebookSource.pm [hunk 2]
```diff
@@ -25,44 +27,17 @@ has 'redirect_url' => (isa => 'Str', is => 'rw', required => 1, default => 'http
...
-    my $username =  $params->{'username'};
-    foreach my $condition (@{ $own_conditions }) {
-        if ($condition->{'attribute'} eq "username") {
-            if ( $condition->matches("username", $username) ) {
-                push(@{ $matching_conditions }, $condition);
-            }
-        }
-    }
-    return $username;
+    person_modify( $pid, firstname => $info->{first_name}, lastname => $info->{last_name} );
```
Comment (fdurand) [inline lib/pf/Authentication/Source/FacebookSource.pm:39]: I suggest to add the portal and the source_id as param (we use that to instantiate the correct portal profile on  the portal on the management interface)

File: lib/pf/Authentication/Source/FacebookSource.pm [hunk 3]
```diff
@@ -25,44 +27,17 @@ has 'redirect_url' => (isa => 'Str', is => 'rw', required => 1, default => 'http
...
-    my $self = shift;
-    return([@{$self->SUPER::available_attributes}, {value => 'username', type => $Conditions::SUBSTRING }]);
-}
-
-=head2 match_in_subclass
-
-=cut
+sub lookup_from_provider_info {
+    my ( $self, $pid, $info ) = @_;
+    my $logger = get_logger();
```
Comment (fdurand) [inline lib/pf/Authentication/Source/FacebookSource.pm]: useless

File: lib/pf/Authentication/Source/GithubSource.pm [hunk 4]
```diff
@@ -22,47 +24,20 @@ has 'access_token_path' => (isa => 'Str', is => 'rw', default => '/login/oauth/a
...
-                push(@{ $matching_conditions }, $condition);
-            }
-        }
-    }
-    return $username;
+sub lookup_from_provider_info {
+    my ( $self, $pid, $info ) = @_;
+    my $logger = get_logger();
+    my ($first_name, $last_name) = split(' ', $info->{name});
+    person_modify( $pid, firstname => $first_name, lastname => $last_name );
```
Comment (fdurand) [inline lib/pf/Authentication/Source/GithubSource.pm:38]: same remark (portal and the source_id as param)

File: lib/pf/Authentication/Source/GoogleSource.pm [hunk 5]
```diff
@@ -26,47 +28,19 @@ has 'redirect_url' => (isa => 'Str', is => 'rw', required => 1, default => 'http
...
-    my $username =  $params->{'username'};
-    foreach my $condition (@{ $own_conditions }) {
-        if ($condition->{'attribute'} eq "username") {
-            if ( $condition->matches("username", $username) ) {
-                push(@{ $matching_conditions }, $condition);
-            }
-        }
-    }
-    return $username;
+    person_modify( $pid, firstname => $info->{given_name}, lastname => $info->{family_name} );
```
Comment (fdurand) [inline lib/pf/Authentication/Source/GoogleSource.pm:38]: Same remark (portal and the source_id as param)

File: lib/pf/Authentication/Source/WindowsLiveSource.pm [hunk 6]
```diff
@@ -26,44 +28,17 @@ has 'redirect_url' => (isa => 'Str', is => 'rw', required => 1, default => 'http
...
-    my $username =  $params->{'username'};
-    foreach my $condition (@{ $own_conditions }) {
-        if ($condition->{'attribute'} eq "username") {
-            if ( $condition->matches("username", $username) ) {
-                push(@{ $matching_conditions }, $condition);
-            }
-        }
-    }
-    return $username;
+    person_modify( $pid, firstname => $info->{first_name}, lastname => $info->{last_name} );
```
Comment (fdurand) [inline lib/pf/Authentication/Source/WindowsLiveSource.pm:38]: Same remark (portal and the source_id as param)

File: lib/pf/Authentication/Source/WindowsLiveSource.pm [hunk 7]
```diff
@@ -26,44 +28,17 @@ has 'redirect_url' => (isa => 'Str', is => 'rw', required => 1, default => 'http
...
-    my $self = shift;
-    return([@{$self->SUPER::available_attributes}, {value => 'username', type => $Conditions::SUBSTRING }]);
-}
-
-=head2 match_in_subclass
-
-=cut
+sub lookup_from_provider_info {
+    my ( $self, $pid, $info ) = @_;
+    my $logger = get_logger();
```
Comment (fdurand) [inline lib/pf/Authentication/Source/WindowsLiveSource.pm]: useless

File: lib/pf/Authentication/Source/GoogleSource.pm [hunk 8]
```diff
@@ -26,47 +28,19 @@ has 'redirect_url' => (isa => 'Str', is => 'rw', required => 1, default => 'http
...
 
-sub available_attributes {
-    my $self = shift;
-    return([@{$self->SUPER::available_attributes}, {value => 'username', type => $Conditions::SUBSTRING }]);
-}
-
-=head2 match_in_subclass
+sub lookup_from_provider_info {
+    my ( $self, $pid, $info ) = @_;
+    my $logger = get_logger();
```
Comment (fdurand) [inline lib/pf/Authentication/Source/GoogleSource.pm]: useless

File: lib/pf/Authentication/Source/GithubSource.pm [hunk 9]
```diff
@@ -22,47 +24,20 @@ has 'access_token_path' => (isa => 'Str', is => 'rw', default => '/login/oauth/a
...
-        if ($condition->{'attribute'} eq "username") {
-            if ( $condition->matches("username", $username) ) {
-                push(@{ $matching_conditions }, $condition);
-            }
-        }
-    }
-    return $username;
+sub lookup_from_provider_info {
+    my ( $self, $pid, $info ) = @_;
+    my $logger = get_logger();
```
Comment (fdurand) [inline lib/pf/Authentication/Source/GithubSource.pm]: useless

File: lib/pf/Authentication/Source/FacebookSource.pm (hunk 2 above)
Comment (julsemaan) [inline lib/pf/Authentication/Source/FacebookSource.pm:39]: The person modify is still there in the Catalyst controllers. This is only to add information.

The info only contains the JSON response from the oauth provider

File: lib/pf/Authentication/Source/WindowsLiveSource.pm (hunk 7 above)
Comment (julsemaan) [inline lib/pf/Authentication/Source/WindowsLiveSource.pm]: I'll remove those although I find it useful sometimes to have this declared everywhere so if I need to add a logging statement quickly, I don't have to do the import + declaration.

**Pull request merged**